specs/agents/metrics-health.md

```diff
@@ -0,0 +1,98 @@
+## Agent Health Metrics
+
+Agents SHOULD expose metrics which allow the analysis of the internal state of the agent, especially including the [reporting/transport mechanism](transport.md).
+
+Many of these metrics are `COUNTER`s: a `COUNTER` is a monotonically increasing metric.
+It is periodically reported by sending the difference of the counter-value now vs its value at the time of the previous report.
+In order to save elasticsearch disk space, zero-increases are not sent.
+
+Example:
+| **Time** | **Counter** | **Reported Metric Value** |
+|----------|-------------|---------------------------|
+| 0s       | 7           | 7                         |
+| 30s      | 10          | 3                         |
+| 1m       | 10          | 0 (not sent)              |
+| 1m30s    | 12          | 2                         |
+
+
+### Event Count Metrics
+
+Agents send their data to the Intake API of the APM-server in the form of a stream of events (e.g. transactions, spans, metricsets...).
+
+The following metrics counting these events SHOULD be exposed:
+ - `agent.events.total`: `COUNTER` of events either submitted to the internal reporting queue or dropped upon submit attempt
+ - `agent.events.dropped.queue`: `COUNTER` of events where the attempt to add it to the reporting queue failed (e.g. due to a full queue)
+ - `agent.events.dropped.error`: `COUNTER` of events dropped after they previously have been successfully added to the queue.
+
+If an agent supports these metrics, the metrics MUST be captured for all events which would reach the APM-Server in an ideal, failure-free world.
+Therefore an event MUST be taken into account for these metrics if an attempt is made to submit it to the internal reporting queue, regardless of the success.
+Events dropped before this attempt (e.g. due to [sampling](tracing-sampling.md) or [transaction_max_spans](handling-huge-traces/tracing-spans-limit.md)) are NOT counted.
+
+The value of `agent.events.dropped.error` is an upper bound for the actual number of dropped events after entering the queue.
+For example, when an request to the Intake API fails without a response, all events within this request MUST be considered as failed.
+In reality, it is possible that e.g. half of the data was actually successfully received and forwarded by the APM server.
+When the APM server responds with an HTTP error code, the number of dropped events SHOULD be computed by subtracting the `accepted` field from the response body from the total number of inflight events of this request.
+
+The `agent.events.total` metric MUST have the label `eventType`. The possible values for `eventType` are the lowercase names of the events according to the [spec](https://github.com/elastic/apm-server/tree/main/docs/spec/v2) (e.g. `transaction`, `metricset`) or the value `other`.
+The `agent.events.dropped.*` metrics MUST NOT have any labels.
```

Comment:
I tried to design the metrics based on the questions they are supposed to answer when looking at them in a dashboard.

For the `agent.events.dropped.*` metrics I have the following questions in mind:
 * Did I not loose any data on the agent side? (sum of `agent.events.dropped.*` is zero)
 * How much data did I loose at maximum (pessimistic upper bound)? (sum of `agent.events.dropped.*`)
 * Was the data lost due to queue overload or other errors (network etc)? (`dropped.queue` vs `dropped.error`)
 
If both `agent.events.dropped.*` are labeled with the `eventType`, then it is possible to answer these questions for individual types of events. E.g. "I know I lost some data, but did I at least not loose any `metricsets`, so I can trust my metrics data? ".

If we now add the label `eventType` only to `dropped.queue` and not to `dropped.error`, I don't see any new valuable questions we could answer. The only one I can come up with is "How many transactions/metricsets did I loose due to the queue being full"? But I don't see any actual insight gained by this, e.g. If you didn't loose any metrics due to the queue being full it does not imply that you didn't loose any metrics at all and therefore you still can't really trust your metric data.

So, event though it is technically possible (and maybe easy) I still wouldn't capture the `eventType` for `dropped.queue`, because imo it doesn't provide more insights to the user viewing these metrics.
This is just my opinion, feel free to disagree here!

Reply:
I suppose let's just start without it, as you say; we can always add it later. Summing over all the types would be equivalent.

Reply:
> One data point: in the Go Agent we have a separate queue/buffer for metrics.

Well that changes things up a bit for the queue utilization metrics. For which queue would you report `agent.events.queue.min_size.pct` and `agent.events.queue.max_size.pct`?
Options I see:
 * Add labels and expose the utilization per queue
 * Always use the non-metrics queue
   * Imagine an app which collects no traces but high cardinality metrics: metricsets might be dropped but the `max_size.pct` would remain at zero.
 * Compute MAX and MIN across the individual queues
   * Would imo render `min_size` useless, because the metrics queue very likely is almost always empty
 * Report the (min,max) utitlization of the queue, which has the highest max-usage in the current reporting time frame. Use a deterministic order if max-usage is equal across the queues
   * Imo the best solution here
   
Whichever solution we chose (might be none of the above), I think we should add it to the spec.

Reply:
To elaborate a bit further after additional discussions at the offsite:

> Add labels and expose the utilization per queue

This is also not really feasible in my opinion.
First of all, I think that the metrics should reflect the knobs that the user is able to turn on the agents.
I think we have a single knob for tuning all of the queues at once in most agents (e.g. `ELASTIC_APM_API_BUFFER_SIZE` for the go agent, @axw please correct me if I'm wrong). Of course we as agent developers are interested in which specific queue is full. But we will use logs to get this information, not the metrics.

In addition, depending on the chosen labels, it might be hard to create a generic Kibana dashboard for all agent languages.
If we decide to add a label `queue-name`, where the label values are defined by each agent implementation, the only option (to my knowledge) in Kibana would be to display the utilization of all queues in a single chart via breakdown. However, this would be quiet confusing in comparison to having one char per queue I think.

Alternatively we could use `eventType` (span, transaction, etc) as label. This has significant downsides in case of queues shared for multiple event types, e.g. in the Java Agent.
For example, if the shared queue on the java agent is full, the `max_usage` metric would be on 100% for all `eventTypes`, even though it might be filled only with spans. When now looking at the `max_usage` for `metricsets`, this might lead users to the wrong conclusion that we need to reduce the number of metrics reported.

This is why I would propose this solution from my last comment as solution:
> Report the (min,max) utitlization of the queue, which has the highest max-usage in the current reporting time frame. Use a deterministic order if max-usage is equal across the queues

@axw @felixbarny Any thoughts / objections?

Reply:
> If we decide to add a label queue-name, where the label values are defined by each agent implementation, the only option (to my knowledge) in Kibana would be to display the utilization of all queues in a single chart via breakdown. 

I think we could visualize it by showing the max utilization of all queues which would be similar to 

> Report the (min,max) utitlization of the queue, which has the highest max-usage in the current reporting time frame. 

With the difference that we do this at the UI layer. This has the advantage that we can at least potentially drill in deeper (even if we don't have a dedicated UI for it in the first iteration).

I think the approach to report a utilization metric per queue gives us more optionality going forward with little impact on storage and complexity. But maybe I'm missing something.

Reply:
>I think we could visualize it by showing the max utilization of all queues which would be similar to
>>Report the (min,max) utitlization of the queue, which has the highest max-usage in the current reporting time frame.

Regarding the `max_usage` this would work properly. However we would also get the `min_usage` across all queues instead of the queue which is currently under load. Therefore it is rather equal to this approach, but on visualization side:

> Compute MAX and MIN across the individual queues
Would imo render min_size useless, because the metrics queue very likely is almost always empty

I am okay with that as this would merely be a visualization limitation currently in kibana and we could always do a per-language dashboard with separate graphs per queue.

If you agree, I will add the optional `queueName` label to the spec with implementation-defined values per agent.

Reply:
Yes, snake_case should be preferred. I think some agents may have used camelCase, but it's not the norm.